app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt

```diff
@@ -54,152 +48,50 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
-              showDivider
-            )
-          itemList.add(olderSectionTitleViewModel)
-          for (promotedStory in it.promotedStoryList.olderPlayedStoryList) {
-            val ongoingStoryViewModel =
-              OngoingStoryViewModel(
-                promotedStory,
-                entityType,
-                fragment as OngoingStoryClickListener
-              )
-            itemList.add(ongoingStoryViewModel)
-          }
-        }
-
-        if (it.promotedStoryList.suggestedStoryList.isNotEmpty()) {
-          val showDivider = itemList.isNotEmpty()
-          val recommendedSectionTitleViewModel =
-            SectionTitleViewModel(
-              activity.getString(R.string.recommended_stories),
-              showDivider
-            )
-          itemList.add(recommendedSectionTitleViewModel)
-          for (suggestedStory in it.promotedStoryList.suggestedStoryList) {
-            val ongoingStoryViewModel =
-              OngoingStoryViewModel(
-                suggestedStory,
-                entityType,
-                fragment as OngoingStoryClickListener
-              )
-            itemList.add(ongoingStoryViewModel)
-          }
+  private fun createRecyclerViewAdapter(): BindableAdapter<RecentlyPlayedItemViewModel> {
+    return BindableAdapter.MultiTypeBuilder
+      .newBuilder<RecentlyPlayedItemViewModel, ViewType> { viewModel ->
+        when (viewModel) {
+          is SectionTitleViewModel -> ViewType.SECTION_TITLE_TEXT
+          is OngoingStoryViewModel -> ViewType.SECTION_STORY_ITEM
+          else -> throw IllegalArgumentException("Encountered unexpected view model: $viewModel")
         }
-
-        binding.ongoingStoryRecyclerView.layoutManager =
-          createLayoutManager(
-            it.promotedStoryList.recentlyPlayedStoryCount,
-            it.promotedStoryList.olderPlayedStoryCount,
-            it.promotedStoryList.suggestedStoryCount
-          )
-        ongoingListAdapter.notifyDataSetChanged()
       }
-    )
-  }
-
-  private fun getAssumedSuccessfulPromotedActivityList(): LiveData<PromotedActivityList> {
-    // If there's an error loading the data, assume the default.
-    return Transformations.map(ongoingStoryListSummaryResultLiveData) {
-      it.getOrDefault(
-        PromotedActivityList.getDefaultInstance()
+      .setLifecycleOwner(fragment)
+      .registerViewDataBinder(
+        viewType = ViewType.SECTION_TITLE_TEXT,
+        inflateDataBinding = SectionTitleBinding::inflate,
+        setViewModel = SectionTitleBinding::setViewModel,
+        transformViewModel = { it as SectionTitleViewModel }
       )
-    }
+      .registerViewDataBinder(
+        viewType = ViewType.SECTION_STORY_ITEM,
+        inflateDataBinding = OngoingStoryCardBinding::inflate,
+        setViewModel = OngoingStoryCardBinding::setViewModel,
+        transformViewModel = { it as OngoingStoryViewModel }
+      )
+      .build()
   }
 
-  private fun createLayoutManager(
```

Comment:
Not able to call this correctly.

Reply:
I took a look at this and it looks like this layout manager logic is only so that the spanCount can be instantiated properly for the GridLayout manager. I suggest it might be a better design to follow what was done in the [HomeFragmentPresenter](https://github.com/oppia/oppia-android/blob/develop/app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt#L73) and [HomeViewModel](https://github.com/oppia/oppia-android/blob/develop/app/src/main/java/org/oppia/android/app/home/HomeViewModel.kt#L41). The `HomeViewModel` has a list of view models in the `HomeFragment`'s recycler view; this list of view models are view models for each section shown in the home fragment (e.g. WelcomeViewModel, PromotedStoriesViewModel, etc.) that have layouts and presenters that are individually managed.

For this PR, it would mean creating a few more view models:
1. A `RecentlyPlayedViewModel` for the `RecentlyPlayedFragment` that contains a list of view models.
2. This list would contain a new model (maybe something called `RecentlyPlayedSectionViewModel`) that represents each section we want to show in this fragment.
Each `RecentlyPlayedSectionViewModel` would then consist of a `SectionTitleViewModel` and a list of `OngoingStoryViewModel`s for the stories shown in that individual section (this is what the recently played fragment's list is).

The recently played fragment then only has to manage a list of the sections it wants to display (instead of offsetting the position to hide them) so then individual sections manage its own grid layout and can more easily set its span count its own list of stories. Any sections not shown in the `RecentlyPlayedFragment` just wouldn't have a view model in the `RecentlyPlayedFragmentViewModel`'s list. A caveat, though, is that we might have to create a new fragment/presenter for each `RecentlyPlayedSection` that is managed by the `RecentlyPlayedSectionViewModel`.

Does this make sense?

cc/ @rt4914 WDYT?

Reply:
Sounds good to me.